### ISSUE TITLE
Contact IP address erroneous in 180 ringing and 200 OK

### ISSUE THREAD

**Issue by JuanPascual2** [alfredh/baresip#431]
Hi
In the windows version with a tls configuration, Baresip sends in the 180 Ringing and 200 OK the contact: Contact<sip:baresipjuan1-0x2275d50@0.0.0.0:49685;transport=tls>
0.0.0.0 isn't correct IP address.
In the linux version, the contact is OK.
It contains the IP address of the pc.
BR
Juan

**Comment by alfredh** [comment]
hi @JuanPascual2

can you please include a full debug log from baresip?

and also the output of the `/sipstat` command ?


Alfred

**Comment by JuanPascual2** [comment]
Hello
I ran barsip with option  -v but  i don't find any log file.
 Is it the good option  to debug log?

The result sipstat is:
baresipjuan1@vlt.com: {0/TLS/v4} 200 OK (VLT Orange) [1 binding]
←[32mAll 1 useragent registered successfully! (12 ms)←[;m
call: alloc with params laddr=10.155.139.214, af=AF_INET, use_rtp=1
call: use_video=0
answering call from sip:+33960801717@vlt.com with 200
call: update media
audio: Set audio decoder: PCMA 8000Hz 1ch
audio: player started with sample format S16LE
audio: Set audio encoder: PCMA 8000Hz 1ch
audio: source started with sample format S16LE
audio tx pipeline:     winwave ---> vumeter ---> PCMA
audio rx pipeline:     winwave <--- vumeter <--- PCMA
main: long async blocking: 280>100 ms (h=0x43fd40 arg=0xa87e58)
transports:[                ]←[;m
  10.155.139.214:62174 (UDP)
  10.155.139.214:56581 (TCP)
  10.155.139.214:56582 (TLS)
client transactions:
server transactions:
  INVITE     ACCEPTED   18s (z9hG4bKb777.4a42ad16000000000000000000000000.0)
transports:[=               ]←[;m
  10.155.139.214:62174 (UDP)
  10.155.139.214:56581 (TCP)
  10.155.139.214:56582 (TLS)
client transactions:
server transactions:
  INVITE     ACCEPTED   10s (z9hG4bKb777.4a42ad16000000000000000000000000.0)
sip:+33960801717@vlt.com: session closed: unknown error
tmr: long async blocking: 125>100 ms (h=0x42c6d5 arg=0xa8b748)

I tested with kamailio below kamailio log.

Request Kamailio

INVITE sip:baresipjuan1@vlt.com;transport=tls SIP/2.0
Via: SIP/2.0/TLS 10.155.139.170:5081;branch=z9hG4bKb777.4a42ad16000000000000000000000000.0
To: <sip:baresipjuan1@vlt.com>
From: <sip:+33960801717@vlt.com>;tag=234512;tag=c544751ca238ee22e127ad45c3682d75-d361
CSeq: 130472 INVITE
Call-ID: 26f374903a346171-23068@10.155.139.170
Max-Forwards: 70
Content-Length: 138
Contact:<sip:vlt@10.155.139.170:5081>
Content-Type: application/sdp
Voipm:2222

v=0
o=root 1807992530 1807992530 IN IP4 192.168.1.1
s=VLT
c=IN IP4 192.168.1.1
t=0 0
m=audio 31814 RTP/AVP 8
a=rtpmap:8 PCMA/8000


Response Baresip

SIP/2.0 180 Ringing
Via: SIP/2.0/TLS 10.155.139.170:5081;branch=z9hG4bKb777.4a42ad16000000000000000000000000.0
To: <sip:baresipjuan1@vlt.com>;tag=703f37b9d5e6bbdb
From: <sip:+33960801717@vlt.com>;tag=234512;tag=c544751ca238ee22e127ad45c3682d75-d361
CSeq: 130472 INVITE
Call-ID: 26f374903a346171-23068@10.155.139.170
Server: baresip v0.5.9 (i686/win32)
Contact: <sip:baresipjuan1-0xa86bc8@0.0.0.0:56585;transport=tls>
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,REFER,NOTIFY,SUBSCRIBE,INFO,MESSAGE
Content-Length: 0

SIP/2.0 200 Answering
Via: SIP/2.0/TLS 10.155.139.170:5081;branch=z9hG4bKb777.4a42ad16000000000000000000000000.0
To: <sip:baresipjuan1@vlt.com>;tag=703f37b9d5e6bbdb
From: <sip:+33960801717@vlt.com>;tag=234512;tag=c544751ca238ee22e127ad45c3682d75-d361
CSeq: 130472 INVITE
Call-ID: 26f374903a346171-23068@10.155.139.170
Server: baresip v0.5.9 (i686/win32)
Contact: <sip:baresipjuan1-0xa86bc8@0.0.0.0:56585;transport=tls>
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,REFER,NOTIFY,SUBSCRIBE,INFO,MESSAGE
Content-Type: application/sdp
Content-Length: 257

v=0
o=- 1002100018 926630515 IN IP4 10.155.139.214
s=-
c=IN IP4 10.155.139.214
t=0 0
a=tool:baresip 0.5.9
m=audio 7820 RTP/AVP 8
a=rtpmap:8 PCMA/8000
a=sendrecv
a=label:1
a=rtcp-rsize
a=ssrc:1482193259 cname:sip:baresipjuan1@vlt.com
a=ptime:20

**Comment by alfredh** [comment]
hi @JuanPascual2 

thanks for the logs. the Baresip output and log is the same :)

could you please try this patch for libre and send the log again:

https://gist.github.com/alfredh/2a47125c06b3ca43326696e7d7e9ad01

1. get a clean libre from GitHub
2. apply the patch
3. recompile libre
4. recompile Baresip

Run the same scenario, and paste the Baresip output here.

BTW: how are you compiling Baresip (Visual Studio, mingw) ?

Alfred

**Comment by JuanPascual2** [comment]
Hi Alfred
I compile with mingw.
I applied the patch.
Below log, I still have the address "0.0.0.0"
Juan

baresip.exe -f configkamailio -v
baresip v0.5.9 Copyright (C) 2010 - 2018 Alfred E. Heggestad et al.
Local network address:  IPv4=:10.172.126.228
ui: wincons
aucodec: PCMU/8000/1
aucodec: PCMA/8000/1
aufilt: vumeter
winwave: output devices: 1, input devices: 1
ausrc: winwave
auplay: winwave
dll: open: ./dshow.dll LoadLibraryA() failed
←[31mmodule dshow.dll: unknown error
←[;mmedianat: stun
medianat: turn
medianat: ice
mediaenc: srtp
mediaenc: srtp-mand
mediaenc: srtp-mandf
mediaenc: dtls_srtpf
mediaenc: dtls_srtp
mediaenc: srtp-mandf
DTLS-SRTP ready with profiles SRTP_AES128_CM_SHA1_80:SRTP_AES128_CM_SHA1_32
uuid: loaded UUID f1ce8b4b-7e52-eac2-2996-f5c18dc95fdf from file configkamailio/
uuid
sip: connect: local addr 0.0.0.0:54991
Populated 1 account
module: loading app auloop.dll
module: loading app contact.dll
Populated 3 contacts
module: loading app debug_cmd.dll
module: loading app menu.dll
module: loading app vidloop.dll
Populated 2 audio codecs
Populated 1 audio filter
Populated 0 video codecs
Populated 0 video filters
baresip is ready.
sip: establish: local addr 10.172.126.228:54991
baresipjuan1@vlt.com: {0/TLS/v4} 200 OK (VLT Orange) [1 binding]
←[32mAll 1 useragent registered successfully! (85 ms)←[;m
transports:
  10.172.126.228:56909 (UDP)
  10.172.126.228:54987 (TCP)
  10.172.126.228:54988 (TLS)
client transactions:
server transactions:
call: alloc with params laddr=10.172.126.228, af=AF_INET, use_rtp=1
call: use_video=0
answering call from sip:+33960801717@vlt.com with 200
call: update media
audio: Set audio decoder: PCMA 8000Hz 1ch
audio: player started with sample format S16LE
audio: Set audio encoder: PCMA 8000Hz 1ch
audio: source started with sample format S16LE
audio tx pipeline:     winwave ---> vumeter ---> PCMA
audio rx pipeline:     winwave <--- vumeter <--- PCMA
main: long async blocking: 629>100 ms (h=0x43fd88 arg=0x22d7578)
transports:[                ]←[;m
  10.172.126.228:56909 (UDP)
  10.172.126.228:54987 (TCP)
  10.172.126.228:54988 (TLS)
client transactions:
server transactions:
  INVITE     ACCEPTED   14s (z9hG4bK1693.799c7b84000000000000000000000000.0)
sip:+33960801717@vlt.com: session closed: unknown error
tmr: long async blocking: 147>100 ms (h=0x42c71d arg=0x22f6f38)
sip: connect: local addr 0.0.0.0:55022
sip: establish: local addr 10.172.126.228:55022
 
 Invite sending by Kamailio
 
INVITE sip:baresipjuan1@vlt.com;transport=tls SIP/2.0
Via: SIP/2.0/TLS 10.155.139.170:5081;branch=z9hG4bK1693.799c7b84000000000000000000000000.0
To: <sip:baresipjuan1@vlt.com>
From: <sip:+33960801717@vlt.com>;tag=234512;tag=c544751ca238ee22e127ad45c3682d75-1f58
CSeq: 130472 INVITE
Call-ID: 4363aee3507aa224-27941@10.155.139.170
Max-Forwards: 70
Content-Length: 138
User-Agent: VLT Orange
Contact:<sip:vlt@10.155.139.170:5081>
Content-Type: application/sdp
Voipm:2222

v=0
o=root 1807992530 1807992530 IN IP4 192.168.1.1
s=VLT
c=IN IP4 192.168.1.1
t=0 0
m=audio 31814 RTP/AVP 8
a=rtpmap:8 PCMA/8000

Response Baresip

SIP/2.0 180 Ringing
Via: SIP/2.0/TLS 10.155.139.170:5081;branch=z9hG4bK1693.799c7b84000000000000000000000000.0
To: <sip:baresipjuan1@vlt.com>;tag=4409260a44b4c72d
From: <sip:+33960801717@vlt.com>;tag=234512;tag=c544751ca238ee22e127ad45c3682d75-1f58
CSeq: 130472 INVITE
Call-ID: 4363aee3507aa224-27941@10.155.139.170
Server: baresip v0.5.9 (i686/win32)
Contact: <sip:baresipjuan1-0x22d8e80@0.0.0.0:54991;transport=tls>
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,REFER,NOTIFY,SUBSCRIBE,INFO,MESSAGE
Content-Length: 0

SIP/2.0 200 Answering
Via: SIP/2.0/TLS 10.155.139.170:5081;branch=z9hG4bK1693.799c7b84000000000000000000000000.0
To: <sip:baresipjuan1@vlt.com>;tag=4409260a44b4c72d
From: <sip:+33960801717@vlt.com>;tag=234512;tag=c544751ca238ee22e127ad45c3682d75-1f58
CSeq: 130472 INVITE
Call-ID: 4363aee3507aa224-27941@10.155.139.170
Server: baresip v0.5.9 (i686/win32)
Contact: <sip:baresipjuan1-0x22d8e80@0.0.0.0:54991;transport=tls>
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,REFER,NOTIFY,SUBSCRIBE,INFO,MESSAGE
Content-Type: application/sdp
Content-Length: 257

v=0
o=- 3378883869 817432110 IN IP4 10.172.126.228
s=-
c=IN IP4 10.172.126.228
t=0 0
a=tool:baresip 0.5.9
m=audio 46934 RTP/AVP 8
a=rtpmap:8 PCMA/8000
a=sendrecv
a=label:1
a=rtcp-rsize
a=ssrc:495706699 cname:sip:baresipjuan1@vlt.com
a=ptime:20

**Comment by alfredh** [comment]
thanks! now, can you please try again with this patch:

```
diff --git a/src/sip/transp.c b/src/sip/transp.c
index dd3b7c4..3546c20 100644
--- a/src/sip/transp.c
+++ b/src/sip/transp.c
@@ -459,6 +459,8 @@ static void tcp_estab_handler(void *arg)
        struct le *le;
        int err;
 
+       tcp_conn_local_get(conn->tc, &conn->laddr);
+
        conn->established = true;
 
        le = list_head(&conn->ql);
```

(only this patch on top of git head)

it looks like `tcp_conn_local_get` is not working properly on Windows.
when called right after `tcp_connect` the returned address is 0.0.0.0
when called in the established handler, the address is correct.

**Comment by JuanPascual2** [comment]
Hi
This patch fixes the problem, the contact contains now the IP address.
I use Windows 7.
Thanks
Juan

**Comment by alfredh** [comment]
hi @JuanPascual2 

thanks for report and testing the fix. We will try to make a proper fix in libre.

**Comment by alfredh** [comment]
this was fixed in libre commit 4ded6cbcc9f5a9375fb885109a002d843a27531d
